azurelinuxagent/ga/exthandlers.py

```diff
@@ -391,61 +395,77 @@ def handle_ext_handlers(self, etag=None):
-        '''
+        """
         Check the status of the extension being handled.
         Wait until it has a terminal state or times out.
         Return True if it is handled successfully. False if not.
-        '''
-        handler_i = ExtHandlerInstance(ext_handler, self.protocol)
-        for ext in ext_handler.properties.extensions:
-            ext_completed, status = handler_i.is_ext_handling_complete(ext)
-
-            # Keep polling for the extension status until it becomes success or times out
-            while not ext_completed and datetime.datetime.utcnow() <= wait_until:
-                time.sleep(5)
-                ext_completed, status = handler_i.is_ext_handling_complete(ext)
-
-            # In case of timeout or terminal error state, we log it and return false
-            # so that the extensions waiting on this one can be skipped processing
-            if datetime.datetime.utcnow() > wait_until:
-                msg = "Extension {0} did not reach a terminal state within the allowed timeout. Last status was {1}".format(
-                    ext.name, status)
-                logger.warn(msg)
-                add_event(AGENT_NAME,
-                          version=CURRENT_VERSION,
-                          op=WALAEventOperation.ExtensionProcessing,
-                          is_success=False,
-                          message=msg)
-                return False
+        """
 
-            if status != ValidHandlerStatus.success:
-                msg = "Extension {0} did not succeed. Status was {1}".format(ext.name, status)
-                logger.warn(msg)
-                add_event(AGENT_NAME,
-                          version=CURRENT_VERSION,
-                          op=WALAEventOperation.ExtensionProcessing,
-                          is_success=False,
-                          message=msg)
-                return False
+        def _report_error_event_and_return_false(error_message):
+            # In case of error, return False so that the processing of dependent extensions can be skipped (fail fast)
+            add_event(AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.ExtensionProcessing,
+                      is_success=False,
+                      message=error_message,
+                      log_event=True)
+            return False
+
+        try:
+            handler_i = ExtHandlerInstance(ext_handler, self.protocol)
+
+            # First check if HandlerStatus was a success, if the Handler failed, no need to poll for extension status
+            handler_status = handler_i.get_handler_status()
+            if handler_status is None:
+                msg = "No HandlerStatus available for Handler: {0}, marking the extension as failed".format(
+                    ext_handler.name)
+                return _report_error_event_and_return_false(msg)
+
+            if handler_status.code != ExtensionErrorCodes.PluginSuccess:
```

Comment:
Here you are checking `handler_status.code` (`ExtensionErrorCodes`), but below you are evaluating against `status` (`ValidHandlerStatus`). Why the discrepancy? Can we be consistent?

Reply:
So the `handler_status` is written by the agent, and we use `ExtensionErrorCodes` to specify the state of the handler. The status on the other hand is written by the extension in the status file and they use String representations of `ValidHandlerStatus` to specify the status of the execution of the extension script. That's the primary reason why you see a discrepancy. `ExtensionErrorCodes` are more verbose (multiple options) while the `ValidHandlerStatus` is not (lesser options). Since both were designed for different features, its hard to make it consistent.

Reply:
So why not check `handler_status != ValidHandlerStatus.success` here?

Reply:
`handler_status` is an object of `ExtHandlerStatus()` class here. But the check below is a basic string comparison (`status_from_status_file == "success"`) (these terminologies can be super confusing lol)

Reply:
Sorry, what I meant was check `handler_status.status != ValidHandlerStatus.success`. I don't see why we need to check against the code first, then later against status if we can do status in both cases.

Reply:
The `handler_status.status` is either `Ready` or `NotReady`. We need to report this back to CRP so we cant use this to check the `ValidHandlerStatus.success`. 

Additionally I personally like the code (int) check more since its more verbose and elaborate. `ValidHandlerStatus` is something that the extensions report so that's limited too. The `code` on the other hand is set by the agent, so it can be set to something more specific. Like for now, say we want to ignore the `HeartBeat` failures reported in the `ExtHandlerStatus` then we can just ignore that code here. It gives us more flexibility here if anything.

At the end of the day, they're just 2 different constructs with similar naming which is probably what's throwing you off.

Reply:
Changed the logic, not relying on HandlerStatus anymore

Reply:
Changed the logic to not use handler status now and return True or False if handler was successfully executed